src/casp/bq_scripts/prepare_for_training.py

```diff
@@ -140,42 +168,59 @@ def __get_prepare_cell_info_filter(
     random_seed_offset: int = 0,
     partition_bin_count: int = 40000,
     partition_size: int = 10,
     filter_by_organism: t.Optional[str] = None,
     filter_by_datasets: t.Optional[t.List[str]] = None,
     filter_by_is_primary_data: t.Optional[bool] = None,
+    obs_columns_to_include: t.Optional[t.List[str]] = None,
 ):
     """
     Randomize cells using farm_fingerprint with an offset so we can have deterministic randomization.  See
     https://towardsdatascience.com/advanced-random-sampling-in-bigquery-sql-7d4483b580bb
 
     Then allocate cells into bins of `extract_bin_size` cells.  The last bin may be much
     smaller than this requested size, as it is the remainder cells
-    """
 
-    join_clause = __get_prepare_cell_info_join(project=project, dataset=dataset, filter_by_datasets=filter_by_datasets)
+    :param client: An instance of BigQuery's client object used to interact with the BigQuery service.
+    :param project: The ID of the Google Cloud project where the BigQuery dataset is hosted.
+    :param dataset: The ID of the dataset in BigQuery where the data is hosted.
+    :param extract_table_prefix: A prefix string for naming tables or for similar purposes.
+    :param extract_bin_size: The size for the bins where cells are allocated.
+    :param random_seed_offset: Optional offset for the farm_fingerprint for deterministic randomization.
+        `Defaults:` ``0``
+    :param partition_bin_count: The count of bins for partitioning.
+        `Default:` ``40000``
+    :param partition_size: The size for the partitions.
+        `Default:` ``10``
+    :param filter_by_organism: Optional filter to specify an organism. If not provided, no filtering occurs.
+    :param filter_by_datasets: Optional filter to specify datasets by their names. If not provided, no filtering occurs.
+    :param filter_by_is_primary_data: Optional filter to determine if data is primary or not.
+    :param obs_columns_to_include: Optional list of columns from `cas_cell_info` table to include in ``adata.obs``.
+        If not provided, no specific columns would be added to ``adata.obs`` apart from `cas_cell_index`.
+        Note: It is required to provide the column names along with the aliases for the tables to which they belong.
+        However, the output extract table would contain only the column names, without any aliases.
+        Example: ``["c.cell_type", "c.donor_id", "c.sex", "i.dataset_id"]``
+    """
+    join_clause = __get_prepare_cell_info_join(
+        project=project, dataset=dataset, filter_by_datasets=filter_by_datasets, obs_columns=obs_columns_to_include
+    )
     where_clause = __get_prepare_cell_info_filter(
-        project=project,
-        dataset=dataset,
         organism=filter_by_organism,
         datasets=filter_by_datasets,
         is_primary_data=filter_by_is_primary_data,
     )
-
+    cas_cell_info_columns = ", ".join([*constants.CAS_CELL_INFO_REQUIRED_COLUMNS, *obs_columns_to_include])
```

Comment:
This is potentially problematic if `obs_columns_to_include` also contains the required ones (`constants.CAS_CELL_INFO_REQUIRED_COLUMNS`). Let's make a helper function that reconciles `constants.CAS_CELL_INFO_REQUIRED_COLUMNS` and `obs_columns_to_include` gracefully, like:
```
def reconcile_obs_columns_to_include_with_required(t.Optional[t.List[str]]) -> str:
     return ", ".join([*constants.CAS_CELL_INFO_REQUIRED_COLUMNS] + [
        extra_column for extra_column in obs_columns_to_include
        if extra_column not in constants.CAS_CELL_INFO_REQUIRED_COLUMNS])
```